cpg-concepts/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/concepts/Extensions.kt

```diff
@@ -0,0 +1,180 @@
+ */
+context(TaggingContext)
+fun <T : Node> Selector<T>.with(builder: BuilderContext<T>.() -> OverlayNode): EachContext<T> {
+    val ctx = EachContext(selector = this, builder = { listOf(builder(it)) })
+    listOfEach += ctx
+    return ctx
+}
+
+/**
+ * with: Specifies a [builder] that creates the actual overlay nodes. It is used to assign multiple
+ * overlay nodes to a single selected "underlying" node.
+ */
+context(TaggingContext)
+fun <T : Node> Selector<T>.withMultiple(
+    builder: BuilderContext<T>.() -> List<OverlayNode>
+): EachContext<T> {
+    val ctx = EachContext(selector = this, builder = builder)
+    listOfEach += ctx
+    return ctx
+}
+
+/**
+ * This class holds the context used in the [tag] DSL function. It basically contains a list of
+ * [EachContext], which represent a call to [each] within the tagging context.
+ */
+data class TaggingContext(val listOfEach: MutableList<EachContext<*>> = mutableListOf()) {
+    fun collect(
+        lattice: NodeToOverlayState,
+        state: NodeToOverlayStateElement,
+        node: Node,
+    ): List<OverlayNode> {
+        val list = mutableListOf<OverlayNode>()
+
+        for (each in listOfEach) {
+            // Check, if the node is assignable to the "each"
+            val overlay = each.collect(lattice, state, node)
+            list += overlay
+        }
+
+        return list
+    }
+}
+
+/**
+ * This class represents the context of an [each] DSL function. It contains the [selector] that is
+ * used to select a node that is brought into the [EachContext] as well as a [builder] that
+ * specifies how and which [OverlayNode] is constructed based on the selected node(s).
+ */
+@Suppress("UNCHECKED_CAST")
+data class EachContext<T : Node>(
+    val selector: Selector<T>,
+    val builder: (BuilderContext<T>) -> List<OverlayNode>,
+) {
+    fun collect(
+        lattice: NodeToOverlayState,
+        state: NodeToOverlayStateElement,
+        node: Node,
+    ): List<OverlayNode> {
+        // Check, if our selector matches
+        return if (selector(node as T)) {
+            val ctx = BuilderContext(lattice, state, node, builder)
+            ctx.build()
+        } else {
+            emptyList()
+        }
+    }
+}
+
+/**
+ * This class holds a context that is passed to [EachContext.builder] during execution of the
+ * [EOGConceptPass]. This is needed because the [builder] might need access to the current [state].
+ */
+data class BuilderContext<T : Node>(
+    var lattice: NodeToOverlayState,
+    var state: NodeToOverlayStateElement,
+    var node: T,
+    var builder: (BuilderContext<T>) -> List<OverlayNode>,
+) {
+    fun build(): List<OverlayNode> {
+        return builder(this)
+    }
+}
+
+/**
+ * A selector that describes a possible selection of a CPG [Node] by the following properties:
+ * - its [KClass] (mandatory, see [klass])
+ * - either its [Node.name] (see [namePredicate]) or any other property (see [predicate])
```

Comment:
Interesting choice to connect the `name` and `predicate` check with a logical OR and not an AND (which I would somehow have used/expected). Since it's documented correctly, it's ok.

Reply:
Hmm good point. Might make more sense to combine with an AND to have something like `each<CallExpression>("foo") { it.bar }`